x-pack/plugins/reporting/server/features.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { DEFAULT_APP_CATEGORIES, type Logger } from '@kbn/core/server';
+import { i18n } from '@kbn/i18n';
+import type { PluginSetupContract as FeaturesPluginSetup } from '@kbn/features-plugin/server';
+
+interface FeatureRegistrationOpts {
+  features: FeaturesPluginSetup;
+  deprecatedRoles: string[] | false;
+  isServerless: boolean;
+  logger: Logger;
+}
+
+/**
+ * If xpack.reporting.roles.enabled === true, register Reporting as a feature
+ * that is controlled by user role names. Also, for Serverless register a
+ * 'shell' Reporting Kibana feature.
+ */
+export function registerFeatures({
+  isServerless,
+  features,
+  deprecatedRoles,
+  logger,
+}: FeatureRegistrationOpts) {
+  // Register a 'shell' feature specifically for Serverless. If granted, it will automatically provide access to
+  // reporting capabilities in other features, such as Discover, Dashboards, and Visualizations. On its own, this
+  // feature doesn't grant any additional privileges.
+  if (isServerless) {
+    features.registerKibanaFeature({
+      id: 'reporting',
+      name: i18n.translate('xpack.reporting.features.reportingFeatureName', {
+        defaultMessage: 'Reporting',
+      }),
+      category: DEFAULT_APP_CATEGORIES.management,
+      app: [],
+      privileges: {
+        all: { savedObject: { all: [], read: [] }, ui: [] },
+        // No read-only mode currently supported
+        read: { disabled: true, savedObject: { all: [], read: [] }, ui: [] },
+      },
+    });
+  }
+
+  if (deprecatedRoles !== false) {
+    // refer to roles.allow configuration (deprecated path)
+    const allowedRoles = ['superuser', ...(deprecatedRoles ?? [])];
+    const privileges = allowedRoles.map((role) => ({
+      requiredClusterPrivileges: [],
+      requiredRoles: [role],
+      ui: [],
+    }));
+
+    // self-register as an elasticsearch feature (deprecated)
+    features.registerElasticsearchFeature({
+      id: 'reporting',
+      catalogue: ['reporting'],
+      management: {
+        insightsAndAlerting: ['reporting'],
+      },
+      privileges,
+    });
+  } else {
+    logger.debug(
+      `Reporting roles configuration is disabled. Please assign access to Reporting use Kibana feature controls for applications.`
+    );
+    // trigger application to register Reporting as a subfeature
+    features.enableReportingUiCapabilities();
```

Comment:
Looks like this line of code would be called when isServerless is true. Is that intended or would that be an issue?

Reply:
@tsullivan is it question for me or @kertal (re: https://github.com/elastic/kibana/pull/180362#discussion_r1591540758)? 🙂 I kept the logic we had here before (this was called regardless of offering type), but I can double-check to confirm if it's needed.

Reply:
Apologies @azasypkin, I meant to ask you.

I'm not sure why I pinged the wrong person in https://github.com/elastic/kibana/pull/180362#discussion_r1591540758. I need to get checked for a 🧠🪱 !

Reply:
So, as I see, it's required to support CSV reporting in Serverless, which I believe we still need. Calling this function adds necessary sub-feature privileges for CSV reporting to Discover, Dashboards, and Visualizations, and adds a reporting management app. We rely on this sub-feature privileges even though we don't explicitly show them in role management UI now.

This flag also enables a sub-feature needed for PDF/PNG reports, which we don't support in Serverless, but I don't see any issues with that since we explicitly disable this specific sub-feature privilege in Serverless with this PR (see `serverless.yaml`).

Does it address your concerns @tsullivan ?

Reply:
Yes it does @azasypkin. The piece that I was missing is that we do rely on those sub-feature privileges.